### DISCUSSION TITLE
How to edit code for different sized displays? e.g. 144 servos in 12 x 12 layout

### DISCUSSION THREAD

**Discussion by aaronse** [aaronse/servo-display#5] [Q&A] [answered]
Hello, I am wondering how I could change the code to where it controls 144 servos with the same movements (12x12 servos). Would I only have to change the driver count and the canvas width and height on the top part of the code or do I have to modify more?

_Originally posted by @SamuA00 in https://github.com/aaronse/ServoPixelDisplay/discussions/3#discussioncomment-1775903_

**Comment by aaronse** [comment, marked as answer]
> Hello, I am wondering how I could change the code to where it controls 144 servos with the same movements (12x12 servos). Would I only have to change the driver count and the canvas width and height on the top part of the code or do I have to modify more?
> 
> _Originally posted by @SamuA00 in [#3 (comment)](https://github.com/aaronse/ServoPixelDisplay/discussions/3#discussioncomment-1775903)_

Hello @SamuA00, yes, you're right, some constants and code will need to be edited for builds with different number of servos and/or different layouts:

- DRIVER_COUNT, should be set to number of PCA9685 drivers used.  Current code assumes each driver is connected to 16 servos.  So, should define as total servo count divided by 16, e.g. 9 for 144 == 12x12 servo build.
- CANVAS_WIDTH, servos per row, think of as display pixel width.  So, 12 for a 12x12 build.
- CANVAS_HEIGHT, number of rows, think of as display pixel height.  So, 12 for a 12x12 build.  
- Adafruit_PWMServoDriver pwmDrivers[] defines a list of PCA9685 servo drivers.  Each PCA9685 driver has a unique I2C address, starting at 0x40 (you solder during build assembly).  The code uses this list to figure out which PCA9685 servo driver is connected to a given servo, enabling the code to send commands to a specific driver.  So, for a 144 servo build, 9 entries would need to be defined, each entry with a unique I2C address based on what was soldered during WTD assembly.  See lines ~260 to ~277 to understand tileset layout and wiring sequence assumed by the current code.

**Reply by SamuA00** [reply]
Thank you very much for explaining. Would the first part of the modified code look like this:


1    #include <Wire.h>
2    #include <Adafruit_PWMServoDriver.h>
3    
4    // Servo driver I2C Address definitions
5    Adafruit_PWMServoDriver pwmDrivers[] =
6    {
7      Adafruit_PWMServoDriver(0x40),
8       Adafruit_PWMServoDriver(0x41),
9       Adafruit_PWMServoDriver(0x42),
10       Adafruit_PWMServoDriver(0x43),
11       Adafruit_PWMServoDriver(0x44),
12        Adafruit_PWMServoDriver(0x45),
13       Adafruit_PWMServoDriver(0x46),
14       Adafruit_PWMServoDriver(0x47),
15       Adafruit_PWMServoDriver(0x48)
16    };
17    
18    #define DRIVER_COUNT 9
19    
20    #define SERVOMIN  150 // This is the 'minimum' pulse length count (out of 4096)
21    #define SERVOMAX  600 // This is the 'maximum' pulse length count (out of 4096)
22    #define USMIN  600    // This is the rounded 'minimum' microsecond length based on the minimum pulse of 150
23    #define USMAX  2400   // This is the rounded 'maximum' microsecond length based on the maximum pulse of 600
24    
25    #define SERVO_FREQ 60 // Analog servos run at ~60 Hz updates.  Try 50 (original 
26                          // sample value) if your servos have a seizure using 60.
27    
28    #define SERVO_COUNT 16
29    
30    #define MIN_SERVO_ANGLE 0
31    #define MAX_SERVO_ANGLE 60
32    #define DEFAULT_SERVO_ANGLE 30
33    
34    // Servo panel layout
35    #define TILESET_WIDTH 4
36    #define TILESET_HEIGHT 4
37    #define CANVAS_WIDTH 12
38    #define CANVAS_HEIGHT 12
39    #define PIXEL_COUNT CANVAS_WIDTH * CANVAS_HEIGHT


So do I just have to modify the first part of the code like I did above? Also I am not sure if I did lines 7 through 15 correctly.

**Reply by aaronse** [reply]
> So do I just have to modify the first part of the code like I did above? Also I am not sure if I did lines 7 through 15 correctly.

You're welcome.  Yep, code looks good to me.  Found it helpful to methodically solder the address pins for a batch of PCA9685 servo drivers.  

![image](https://user-images.githubusercontent.com/16479976/145326636-7839c89c-41dc-408c-9971-24ddb7e6321f.png)

![image](https://user-images.githubusercontent.com/16479976/145326463-2e383959-5e84-463d-a256-4affd5cf8269.png)
